src/hp/derp/client_conn.rs

```diff
@@ -0,0 +1,840 @@
+        // 	c.s.bytesSent.Add(int64(len(packet.contents)))
+        // }
+        // }()
+        let srckey = packet.src;
+        let contents = packet.bytes;
+        if srckey.is_zero() {
+            // TODO: ensure we handle this correctly on the client side
+            write_frame_timeout(
+                &mut self.writer,
+                FRAME_RECV_PACKET,
+                vec![&contents],
+                self.timeout,
+            )
+            .await
+        } else {
+            write_frame_timeout(
+                &mut self.writer,
+                FRAME_RECV_PACKET,
+                vec![srckey.as_bytes(), &contents],
+                self.timeout,
+            )
+            .await
+        }
+    }
+}
+
+/// Responsible for reading frames from the client, parsing the frames, and sending
+/// the content to the correct location.
+///
+/// The `ClientConnReader` can:
+///     - receive a ping and notify the `ClientConnWriter` to write a pong back
+///     to the client
+///     - notify the server to send a packet to another peer on behalf of the client
+///     - note whether the client is `preferred`  TODO: what is this?
+///
+/// If the `ClientConnReader` `can_mesh` (is a trusted mesh peer), it can also:
+///     - tell the server to add the current client as a watcher TODO: what is a watcher?
+///     - tell the server to close a given peer
+///     - tell the server to forward a packet from another peer.
+struct ClientConnReader<R>
+where
+    R: AsyncRead + Unpin + Send + Sync,
+{
+    /// Indicates whether this client can mesh
+    can_mesh: bool,
+    /// Reader we use to read from the client
+    reader: R,
+    /// PublicKey of this client
+    key: PublicKey,
+    /// Pong replies to sent to the client, processed by the [`ClientConnWriter`]
+    send_pong: mpsc::Sender<[u8; 8]>,
+
+    /// Channels used to communicate with the server about actions
+    /// it needs to take on behalf of the client
+    server_channels: ServerChannels,
+
+    /// TODO: what is this?
+    preferred: bool,
+}
+
+impl<R> ClientConnReader<R>
+where
+    R: AsyncRead + Unpin + Send + Sync,
+{
+    async fn run(mut self, done: CancellationToken) -> Result<()> {
+        tokio::select! {
+            biased;
+            res = self.read() => {
+                res
+            }
+            _ = done.cancelled() => {
+                Ok(())
+            }
+        }
+    }
+
+    /// Read off the `reader`, erroring and returning at the first error
+    async fn read(&mut self) -> Result<()> {
+        let mut buf = BytesMut::new();
+        loop {
+            // TODO: return Ok(()) if EOF error
+            let (frame_type, _frame_len) =
+                read_frame(&mut self.reader, MAX_FRAME_SIZE, &mut buf).await?;
+            // TODO: "note client activity", meaning we update the server that the client with this
+            // public key was the last one to receive data
+            match frame_type {
```

Comment:
These methods below should take just ownership over the read content. That avoids clones, and fixes a correctness issue here potentially. 

```rust
let (frame_type, frame_len) =  read_frame(&mut self.reader, MAX_FRAME_SIZE, &mut buf).await?;
// assuming frame_len is how much we read

let frame_bytes = buf.split_to(frame_len);
// frame_bytes can be now passed as owned to the next pieces
// `buf` can be used again
```